Gadgets/MatchingGadgets.h

```diff
@@ -295,7 +300,9 @@ class MaxFillAmountsGadget : public GadgetT
         remainingS_buy(pb, constants, remaining.result(), order.amountS.packed, order.amountB.packed, NUM_BITS_AMOUNT, NUM_BITS_AMOUNT, NUM_BITS_AMOUNT, FMT(prefix, ".remainingS_buy")),
         remainingS(pb, order.buy.packed, remainingS_buy.result(), remaining.result(), FMT(prefix, ".remainingS")),
         fillAmountS(pb, order.balanceSBefore.balance, remainingS.result(), NUM_BITS_AMOUNT, FMT(prefix, ".fillAmountS")),
-        fillAmountB(pb, constants, fillAmountS.result(), order.amountB.packed, order.amountS.packed, NUM_BITS_AMOUNT, NUM_BITS_AMOUNT, NUM_BITS_AMOUNT, FMT(prefix, ".fillAmountB"))
+        fillAmountB(pb, constants, fillAmountS.result(), order.amountB.packed, order.amountS.packed, NUM_BITS_AMOUNT, NUM_BITS_AMOUNT, NUM_BITS_AMOUNT, FMT(prefix, ".fillAmountB")),
+        limitedFillAmountS(pb, fillAmountS.result(), maxFillS.value(), NUM_BITS_AMOUNT, FMT(prefix, ".limitedFillAmountS")),
+        limitedFillAmountB(pb, fillAmountB.result(), maxFillB.value(), NUM_BITS_AMOUNT, FMT(prefix, ".limitedFillAmountB"))
```

Comment:
`limitedFillAmountB` can be scaled down here independently of `limitedFillAmountS` which changes the rate of the order. At this step in the order matching code the rate of the order should be kept, the matching code that follows could change the rate of the the taker order in a way that is beneficial to the taker conforming the rate of the maker order (if the orders are matchable).

So I think here only `maxFillS` should be used to lower `fillAmountS` if necessary (like with `filled`), `fillAmountB` should still be calculated using the rate of the order.

Reply:
Thanks Brechet. You mean like this?
```
fillAmountS(); // calc fillS_A
limitedFillAmountS(); // limitedFillAmountS == min(fillAmountS, maxFillS)
fillAmountB(..., limitedFillAmountS.result(), ...); // use limitedFillAmountS to calculate fillB
limitedFillAmountB(); // limitedFillAmountB = min(fillAmountB, maxFillB)
// maybe a accuracy checking here.
```
I think it's OK.


> limitedFillAmountB can be scaled down here independently 

Yes, it can, but I think it will be rejected by CheckValidGadget, is it?? Actually I think maybe we can skip the whole order matching logic in the circuit if CheckValidGadget is good enough. That somehow means Relay then has more power to not only pick orders to be deal with, and also the fill rate of those orders so that they can optimize/parallelize their execution. Any opinion?

Reply:
Yeah, something like that! `CheckValidGadget` should indeed catch it as well if it would be invalid. 

But yes, if we don't want to enforce anything except the order rate we should be able to just use `CheckValidGadget` pretty much as it is now I think. Filled amount should still be checked, but the balance checking shouldn't even be needed because the `subadd` gadgets for the token transfers will disallow underflows of the balances, but to be safe it may not be a bad idea to add the `min` for the balance anyway.

The other thing we still want/need:
- `fillS_A == fillB_B` and `fillS_B == fillB_A`: this is important because this way we only need to store 2 values on-chain (instead of potentially 4).
- We still need buy/sell orders so we can correctly update the trading history for the user (either by using `fillS_A` or `fillS_B` depending on the type). This is to e.g. to allow market buy and market sell orders.

Should make the code a bit easier for sure (though it's only complicated in the circuit, in python/JS it's like 10 lines of code :).

Reply:
Thanks, I will update it soon. For the next step saving the order matching circuit, we can hold it for a while to see if it is necessary for now as existed circuit fully checks everything.

Two more questions. For the ring matching, How can I tell who the taker is and who the maker is? From circuit, It seems A is the taker if A places a 'buy' order, vice verse, is it??
And Relay guys told me that they always set order.buy = 1, I don't think it was required by the circuit, do you know the reason? I think if all orders come in are buy orders, we don't need to check that flag, and can fix the places inside the ring where maker & taker stand.

Reply:
Just forgot, as you mentioned using maxFillS to calculate fillAmountB, I actually have tried that yesterday, but it seemed that I need to change this function a lot to get the same result just as I did here because I can not simply limit the fillAmountS at the end of the calculation, but re-implement this getMaxFillAmounts() function with maxFillS/B come into calculation at the very beginning. However, in the end, it still gives the same result -- (maxFillS, maxFillB). That's why I did not use maxFillS in the calculation and leave the rate check to the following checking gadget, I think this is a better way as it changes less and achieves same result.

Reply:
> For the ring matching, How can I tell who the taker is and who the maker is?  From circuit, It seems A is the taker if A places a 'buy' order, vice verse, is it??

Currently orderA is always the taker and orderB is always the maker, not matter which order is buy/sell. For the actual matching they are sometimes switched, but that's only a trick to get the correct result. Buy/Sell has no impact on the rate the order is settled for market orders, but it's still important because it dictates if we store fillS or fillB in the trading history.

> And Relay guys told me that they always set order.buy = 1, I don't think it is required by the circuit, do you know the reason? I think if all orders come in are buy orders, we don't need to check that flag, and can fix the places inside the ring where maker & taker stand.

If the order is always a buy order than I guess the relayer doesn't support market selling yet? That's the reason we need buy/sell orders. For a market sell order the order owner will set the price high enough to match the most expensive order, but as the order works it's way down the order book it may be matched with cheaper orders, which allows the order owner of the market sell order to get more tokenB than specified in the order and allows him to keep the surplus in tokenS, but the complete amountS will be sold to fill it 100% (and for buy orders we will fill the order until we get 100% of amountB).
(Small side note: we could also use something like fillS + surplus to update the trading history instead of buy/sell orders, but than we'd have the surplus amount onchain as well).

Reply:
> Just forgot, as you mentioned using maxFillS to calculate fillAmountB, I actually have tried that yesterday, but it seemed that I need to change this function a lot to get the same result just as I did here because I can not simply limit the fillAmountS at the end of the calculation. What I need to do is re-implementing this getMaxFillAmounts() function with maxFillS/B come into the calculation at the very beginning. However, in the end, it still gives the same result -- (maxFillS, maxFillB). That's why I did not use maxFillS in the calculation and leave the rate check to the following checking gadget, I think this is a better way as it changes less and achieves same result.

Yeah, this is fine if we don't want to force the strict rules anyway.